### DISCUSSION TITLE
Restart function - issue with backtracked outputs

### DISCUSSION THREAD

**Discussion by Marcus-Silv** [WAM2layers/WAM2layers#283] [Q&A] [answered]
Hello everyone,
I have tested the WAM2layers by stopping and restarting the model in the middle of the tracked period and found some differences in backtracked outputs compared to a continuous run of the model. 

**Is there a way to run backtracking using the restart function without compromising e_track and tagged_precip values in the first days after restarting the model?** 

Below is an overview of differences that I found, and additional details for the comparisons can be found in the attached pdf file at the end of this post. 

**Period analyzed:** 2000-07-01 to 2000-09-01 

**Approach:**
I tested two scenarios. The first one, called standard mode, is simply running the model in a continuous run, as we normally would, with restart set as false in the .yaml file. In the second scenario, called restart mode, I started backtracking with restart set as false, and then stopped backtracking after 2000-08-01, about half of the period analyzed. Then I changed the following settings in the .yaml file and restarted backtracking: 
Track_end_date: 2000-07-31 (the next day following the backtracking order. In the standard mode, this date was set as 2000-09-01) 
restart: true 

These were the only settings that I have modified before restarting the model. 

**Outputs:** 

- During backtracking, Lost Moisture before the restart ranged from -0.01% to -0.1%, whereas in the first backtracked day after restart it went to -23402% and kept increasing up to -24% in the final backtracked day. In the standard mode, lost moisture in the final backtracked day was only -0.03%. 

- E_track values in the first backtracked day after restart were highly different in magnitude and spatial variation, in the order of 249 times lower compared to values in the standard mode. On the second day after restart, mean and total e_track values became only 3% lower than standard mode, but there were still some noticeable spatial differences, which were not apparent from the third backtracked after restart and onwards. With 15 days after restart, restart mode outputs were still 1% lower than the standard mode, but turned almost equivalent in the final backtracked day. Total e_track over the entire 63 days analyzed for the restart mode was 2.6% lower than the standard mode. 

![first day after restart](https://github.com/WAM2layers/WAM2layers/assets/145511101/44d9fa5a-ed74-4ba1-b6e6-c2507fd3752a)
![etrack_restart_test](https://github.com/WAM2layers/WAM2layers/assets/145511101/12f36888-0db0-4853-8692-accef0127f84)

- Tagged_precip in the first backtracked day after restart was also drastically different in magnitude and space variation, but became equivalent from the second backtracked day after restart to the final one. Total tagged_precip over the 63 days analyzed for the restart mode was about 2.2% lower than the standard mode. 

![taggedprecip_restart_test](https://github.com/WAM2layers/WAM2layers/assets/145511101/d58d0f5b-767a-4a80-ac09-7b5cb3873be4)

[Restart function test from WAM-2layers backtracking.pdf](https://github.com/WAM2layers/WAM2layers/files/13343031/Restart.function.test.from.WAM-2layers.backtracking.pdf)

**Comment by Peter9192** [comment, marked as answer]
Hi @Marcus-Silv welcome and thanks for reporting your findings here! I think you're the first person to really try and use the restart functionality so you're contributions are very valuable for improving the code.

With respect to your first point about the lost moisture, I think this can be explained because the 'progresstracker' that writes these log messages is very simple and it's state its not included in the output. I'm still hoping to make this more robust, but it's not on top of our priority list. 

Your second point is more concerning. It might be due to some silly mismatch in the restart time or something like that, but it would require some further investigation. Please also note that there are some big changes waiting to be merged into the code base (#274, #278). Especially the latter seems related to your issue. And would simplify solving it. 

If you want to have a go at trying to pinpoint the issue yourself, you could try [creating an editable installation](https://wam2layers.readthedocs.io/en/latest/develop.html#development-installation-additional-tooling) from this branch (`git checkout bugfix-time` and then `pip install -e .`). I would expect the issue arises somewhere around here:
https://github.com/WAM2layers/WAM2layers/blob/404914dbfae8eacfbffcacc7befb7f4d4842724d/src/wam2layers/tracking/backtrack.py#L115-L120

**Reply by Marcus-Silv** [reply]
Hi Peter,
Thank you very much for the clarification and instructions. 
I still haven't had the chance to pinpoint the issue by myself. I noticed that you guys launched a new version of the model. Another possibility that I could try is to see if this issue persists in this version.

**Comment by ruudvdent** [comment]
Hi @Marcus-Silv , we haven't been able to look in detail at this issue during our recent sprint week, but we upgraded it to an issue: #309

**Reply by Marcus-Silv** [reply]
Ok, Ruud. Thank you!
